src/user_lists/views.py

```diff
@@ -49,3 +54,116 @@ def destroy(self, request, *args, **kwargs):
+        if parent_list_id:
+            queryset = queryset.filter(
+                parent_list_id=parent_list_id,
+                parent_list__created_by=self.request.user,
+                parent_list__is_removed=False,
+            )
+        return queryset
+
+    def get_serializer_class(self):
+        return ListItemDetailSerializer if self.action in ["retrieve", "list"] else ListItemSerializer
+
+    def _validate_parent_list(self, parent_list):
+        if parent_list.created_by != self.request.user or parent_list.is_removed:
+            raise ValidationError({"parent_list": "List not found or you don't have permission."})
+
+    def _get_or_create_item(self, serializer, created_by):
+        parent_list = serializer.validated_data.get("parent_list")
+        self._validate_parent_list(parent_list)
+        try:
+            item = serializer.save(created_by=created_by)
+            _update_list_timestamp(parent_list, created_by)
+            return item, parent_list
+        except IntegrityError:
+            _handle_integrity_error_item()
+
+    def perform_create(self, serializer):
+        self._get_or_create_item(serializer, self.request.user)
+
+    def perform_update(self, serializer):
+        parent_list = serializer.validated_data.get("parent_list", serializer.instance.parent_list)
+        self._validate_parent_list(parent_list)
+        try:
+            serializer.save(updated_by=self.request.user)
+            _update_list_timestamp(parent_list, self.request.user)
+        except IntegrityError:
+            _handle_integrity_error_item()
+
+    def destroy(self, request, *args, **kwargs):
+        instance = self.get_object()
+        parent_list = instance.parent_list
+        instance.delete()
+        _update_list_timestamp(parent_list, request.user)
+        return Response({"success": True}, status=status.HTTP_200_OK)
+
+    def _find_existing_item(self, parent_list, unified_document):
+        return ListItem.objects.filter(
+            parent_list=parent_list, unified_document=unified_document, is_removed=False
+        ).first()
+
+    def _serialize_item(self, item):
+        try:
+            return ListItemDetailSerializer(item, context={"request": self.request}).data
+        except Exception:
+            return {"id": item.id}
+
+    def _existing_item_response(self, existing_item):
+        return Response(
+            {"error": "Item already in list", "item": self._serialize_item(existing_item)},
+            status=status.HTTP_400_BAD_REQUEST,
+        )
+
+    @action(detail=False, methods=["post"], url_path="add-item-to-list")
+    def add_item_to_list(self, request):
+        serializer = ListItemSerializer(data=request.data, context={"request": request})
+        serializer.is_valid(raise_exception=True)
+        parent_list = serializer.validated_data["parent_list"]
+        unified_document = serializer.validated_data["unified_document"]
```

Comment:
If we are referring to the object, that's fine but if this is just an ID, let's rename